frame/dapps-staking/src/testing_utils.rs

```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
 /// Used to perform claim for stakers with success assertion
 pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartContract<AccountId>) {
     let (claim_era, _) = DappsStaking::staker_info(&claimer, contract_id).claim();
-    let init_state = MemorySnapshot::all(claim_era, contract_id, claimer);
+    let current_era = DappsStaking::current_era();
+
+    let init_state_claim_era = MemorySnapshot::all(claim_era, contract_id, claimer);
+    let init_state_current_era = MemorySnapshot::all(current_era, contract_id, claimer);
 
     // Calculate contract portion of the reward
-    let (_, stakers_joint_reward) =
-        DappsStaking::dev_stakers_split(&init_state.contract_info, &init_state.era_info);
+    let (_, stakers_joint_reward) = DappsStaking::dev_stakers_split(
+        &init_state_claim_era.contract_info,
+        &init_state_claim_era.era_info,
+    );
 
-    let (claim_era, staked) = init_state.staker_info.clone().claim();
+    let (claim_era, staked) = init_state_claim_era.staker_info.clone().claim();
     assert!(claim_era > 0); // Sanity check - if this fails, method is being used incorrectly
 
     // Cannot claim rewards post unregister era, this indicates a bug!
-    if let DAppState::Unregistered(unregistered_era) = init_state.dapp_info.state {
+    if let DAppState::Unregistered(unregistered_era) = init_state_claim_era.dapp_info.state {
         assert!(unregistered_era > claim_era);
     }
 
     let calculated_reward =
-        Perbill::from_rational(staked, init_state.contract_info.total) * stakers_joint_reward;
+        Perbill::from_rational(staked, init_state_claim_era.contract_info.total)
+            * stakers_joint_reward;
     let issuance_before_claim = <TestRuntime as Config>::Currency::total_issuance();
 
     assert_ok!(DappsStaking::claim_staker(
         Origin::signed(claimer),
         contract_id.clone(),
     ));
+
+    let final_state_claim_era = MemorySnapshot::all(claim_era, contract_id, claimer);
+    let final_state_current_era = MemorySnapshot::all(current_era, contract_id, claimer);
+
+    // If there's less than 2 events, it should fail anyway, so panic as acceptable
+    let events = dapps_staking_events();
+    let second_last_event = &events[events.len() - 2];
+
+    if DappsStaking::should_restake_reward(
```

Comment:
Since this is `pub`, you should also probably add a few UTs for it.

Also, this function has become quite large - maybe having separate helper functions for case when no restake is done and when it is would help increase readability.